src/main/scala/is/hail/methods/LinearMixedRegression.scala

```diff
@@ -52,32 +55,68 @@ object LinearMixedRegression {
 
     val covNames = "intercept" +: covExpr
 
-    val filteredKinshipMatrix = if (kinshipMatrix.sampleIds sameElements completeSamples)
-      kinshipMatrix
-    else {
-      val fkm = kinshipMatrix.filterSamples(completeSamplesSet)
-      if (!(fkm.sampleIds sameElements completeSamples))
-        fatal("Array of sample IDs in assoc_vds and array of sample IDs in kinship_matrix (with both filtered to complete " +
-          "samples in assoc_vds) do not agree. This should not happen when kinship_matrix is computed from a filtered version of assoc_vds.")
-      fkm
-    }
-
     val n = y.length
     val k = cov.cols
     val d = n - k - 1
 
     if (d < 1)
-      fatal(s"$n samples and $k ${plural(k, "covariate")} including intercept implies $d degrees of freedom.")
+      fatal(s"lmmreg: $n samples and $k ${plural(k, "covariate")} including intercept implies $d degrees of freedom.")
 
     info(s"lmmreg: running lmmreg on $n samples with $k sample ${plural(k, "covariate")} including intercept...")
 
-    val K = new DenseMatrix[Double](n, n, filteredKinshipMatrix.matrix.toBlockMatrixDense().toLocalMatrix().toArray)
+
+    val (fullU, fullS): (DenseMatrix[Double], DenseVector[Double]) = relationMatrix match {
+      case LDMatrix(indexedRowMatrix, variants, numSamplesUsed) => {
+        val variantSet = variants.toSet
+        val localMatrix = new DenseMatrix[Double](indexedRowMatrix.numRows().toInt,
+          indexedRowMatrix.numCols().toInt,
+          indexedRowMatrix.toBlockMatrixDense().toLocalMatrix().toArray)
+        val eigK = eigSymD(localMatrix)
+        val V = eigK.eigenvectors
+        val S = eigK.eigenvalues.map(d => d * (numSamplesUsed.toDouble / variants.size))
+        val sqrtSInv = S.map(d => 1.0 / math.sqrt(variants.length * d))
+
+        //FIXME Need some sort of requirement / assertation that all of the LDMatrix variants are in the VDS. Otherwise fail  s.
+        val filteredVDS = assocVds.filterVariants((v, _, _) => variantSet(v))
+
+        //Each column is all samples for a variant.
+        val sparkGenotypeMatrix = ToNormalizedIndexedRowMatrix(filteredVDS).toBlockMatrixDense().toLocalMatrix()
+        //FIXME This is an array copy of something potentially very large. (toArray copies)
+        val genotypeMatrix = new DenseMatrix[Double](sparkGenotypeMatrix.numRows, sparkGenotypeMatrix.numCols,
+          sparkGenotypeMatrix.toArray).t
+
+        val CV = (genotypeMatrix * V)
```

Comment:
unnecessary (). Here's where you'll try fast distributed matrix multiply, once dan's PR is in, right? You should also benchmark multiplying the genotypeMatrix as IRM by the local V.

Since this route assumes N > M, you should multiply V(* , ::) * sqrtSInv first since that's M^2 instead of N * M complexity.